### ISSUE TITLE
Non deterministic ordering of .c generated file 

### ISSUE THREAD

**Issue by lesteve** [cython/cython#1553]
Here is a snippet that reproduces the problem. We bumped into this in scikit-learn but if this helps I can spend more time on this and try to reproduce on a stand-alone example:

```bash
git clone https://github.com/scikit-learn/scikit-learn > /dev/null 2>&1
cd scikit-learn
(for i in $(seq 1 10); do
    cython sklearn/neighbors/kd_tree.pyx
    md5sum sklearn/neighbors/kd_tree.c
done) | sort
```

Ouptut:
```
2ee7fa6b55aa7f304e9932aa1c29b7b0  sklearn/neighbors/kd_tree.c
2ee7fa6b55aa7f304e9932aa1c29b7b0  sklearn/neighbors/kd_tree.c
434a64388d440852dbb0131df508d9e4  sklearn/neighbors/kd_tree.c
6612c38036533bd4b9905d5e83e02f2e  sklearn/neighbors/kd_tree.c
6612c38036533bd4b9905d5e83e02f2e  sklearn/neighbors/kd_tree.c
6612c38036533bd4b9905d5e83e02f2e  sklearn/neighbors/kd_tree.c
6612c38036533bd4b9905d5e83e02f2e  sklearn/neighbors/kd_tree.c
a2d16d4dd56e088f6959659a0f03126a  sklearn/neighbors/kd_tree.c
a2d16d4dd56e088f6959659a0f03126a  sklearn/neighbors/kd_tree.c
a2d16d4dd56e088f6959659a0f03126a  sklearn/neighbors/kd_tree.c
```

Looking at a few diffs, the differences are ordering only:

```diff
diff --git a/sklearn/neighbors/kd_tree.c b/tmp/kd_tree.c
index 629734c..00802d5 100644
--- a/sklearn/neighbors/kd_tree.c
+++ b/tmp/kd_tree.c
@@ -2121,9 +2121,6 @@ __pyx_memoryview_copy_new_contig(const __Pyx_memviewslice *from_mvs,
                                  size_t sizeof_dtype, int contig_flag,
                                  int dtype_is_object);
 
-/* CIntFromPy.proto */
-static CYTHON_INLINE Py_intptr_t __Pyx_PyInt_As_Py_intptr_t(PyObject *);
-
 /* TypeInfoToFormat.proto */
 struct __pyx_typeinfo_string {
     char string[3];
@@ -2131,6 +2128,9 @@ struct __pyx_typeinfo_string {
 static struct __pyx_typeinfo_string __Pyx_TypeInfoToFormat(__Pyx_TypeInfo *type);
 
 /* CIntFromPy.proto */
+static CYTHON_INLINE Py_intptr_t __Pyx_PyInt_As_Py_intptr_t(PyObject *);
+
+/* CIntFromPy.proto */
 static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *);
 
 /* CIntFromPy.proto */
@@ -45883,6 +45883,50 @@ no_fail:
     return new_mvs;
 }
 
+/* TypeInfoToFormat */
+              static struct __pyx_typeinfo_string __Pyx_TypeInfoToFormat(__Pyx_TypeInfo *type) {
+    struct __pyx_typeinfo_string result = { {0} };
+    char *buf = (char *) result.string;
+    size_t size = type->size;
+    switch (type->typegroup) {
+        case 'H':
+            *buf = 'c';
+            break;
+        case 'I':
+        case 'U':
+            if (size == 1)
+                *buf = (type->is_unsigned) ? 'B' : 'b';
+            else if (size == 2)
+                *buf = (type->is_unsigned) ? 'H' : 'h';
+            else if (size == 4)
+                *buf = (type->is_unsigned) ? 'I' : 'i';
+            else if (size == 8)
+                *buf = (type->is_unsigned) ? 'Q' : 'q';
+            break;
+        case 'P':
+            *buf = 'P';
+            break;
+        case 'C':
+         {
+            __Pyx_TypeInfo complex_type = *type;
+            complex_type.typegroup = 'R';
+            complex_type.size /= 2;
+            *buf++ = 'Z';
+            *buf = __Pyx_TypeInfoToFormat(&complex_type).string[0];
+            break;
+         }
+        case 'R':
+            if (size == 4)
+                *buf = 'f';
+            else if (size == 8)
+                *buf = 'd';
+            else
+                *buf = 'g';
+            break;
+    }
+    return result;
+}
+
 /* CIntFromPy */
               static CYTHON_INLINE Py_intptr_t __Pyx_PyInt_As_Py_intptr_t(PyObject *x) {
     const Py_intptr_t neg_one = (Py_intptr_t) -1, const_zero = (Py_intptr_t) 0;
@@ -46072,50 +46116,6 @@ raise_neg_overflow:
     return (Py_intptr_t) -1;
 }
 
-/* TypeInfoToFormat */
-              static struct __pyx_typeinfo_string __Pyx_TypeInfoToFormat(__Pyx_TypeInfo *type) {
-    struct __pyx_typeinfo_string result = { {0} };
-    char *buf = (char *) result.string;
-    size_t size = type->size;
-    switch (type->typegroup) {
-        case 'H':
-            *buf = 'c';
-            break;
-        case 'I':
-        case 'U':
-            if (size == 1)
-                *buf = (type->is_unsigned) ? 'B' : 'b';
-            else if (size == 2)
-                *buf = (type->is_unsigned) ? 'H' : 'h';
-            else if (size == 4)
-                *buf = (type->is_unsigned) ? 'I' : 'i';
-            else if (size == 8)
-                *buf = (type->is_unsigned) ? 'Q' : 'q';
-            break;
-        case 'P':
-            *buf = 'P';
-            break;
-        case 'C':
-         {
-            __Pyx_TypeInfo complex_type = *type;
-            complex_type.typegroup = 'R';
-            complex_type.size /= 2;
-            *buf++ = 'Z';
-            *buf = __Pyx_TypeInfoToFormat(&complex_type).string[0];
-            break;
-         }
-        case 'R':
-            if (size == 4)
-                *buf = 'f';
-            else if (size == 8)
-                *buf = 'd';
-            else
-                *buf = 'g';
-            break;
-    }
-    return result;
-}
-
 /* CIntFromPy */
               static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *x) {
     const int neg_one = (int) -1, const_zero = (int) 0;
```

More context: we bumped into this is because while trying to use ccache to speed up the .c files compilation and we were seing that the cache was sometimes not reused.

**Comment by lesteve** [comment]
Note: I was using 0.25.2 but I see a similar behaviour in 0.24.1 and 0.23.5.

**Comment by codewarrior0** [comment]
On 12/14/2016 2:34 AM, Loïc Estève wrote:
>
> Here is a snippet that reproduces the problem. We bumped into this in 
> scikit-learn but if this helps I can spend more time on this and try 
> to reproduce on a stand-alone example:
>
> git clone https://github.com/scikit-learn/scikit-learn>  /dev/null2>&1
> cd  scikit-learn
> (for  i  in  $(seq 1 10);  do
>      cython sklearn/neighbors/kd_tree.pyx
>      md5sum sklearn/neighbors/kd_tree.c
> done)|  sort
>
> Ouptut:
>
> |2ee7fa6b55aa7f304e9932aa1c29b7b0 sklearn/neighbors/kd_tree.c 
> 2ee7fa6b55aa7f304e9932aa1c29b7b0 sklearn/neighbors/kd_tree.c 
> 434a64388d440852dbb0131df508d9e4 sklearn/neighbors/kd_tree.c 
> 6612c38036533bd4b9905d5e83e02f2e sklearn/neighbors/kd_tree.c 
> 6612c38036533bd4b9905d5e83e02f2e sklearn/neighbors/kd_tree.c 
> 6612c38036533bd4b9905d5e83e02f2e sklearn/neighbors/kd_tree.c 
> 6612c38036533bd4b9905d5e83e02f2e sklearn/neighbors/kd_tree.c 
> a2d16d4dd56e088f6959659a0f03126a sklearn/neighbors/kd_tree.c 
> a2d16d4dd56e088f6959659a0f03126a sklearn/neighbors/kd_tree.c 
> a2d16d4dd56e088f6959659a0f03126a sklearn/neighbors/kd_tree.c |
>
> Looking at a few diffs, the differences are ordering only:
>
>

Is this Python 3.3+? If so, I'll bet it has to do with the randomized 
hash seed used to give a different dict ordering on each run of the 
program. Try running it with `PYTHONHASHSEED=1 cython sklearn...` and 
see if it becomes consistent.


-- 
David Vierra
MCEdit, a Minecraft World Editor
http://www.mcedit.net
http://twitter.com/codewarrior0

**Comment by lesteve** [comment]
> Try running it with `PYTHONHASHSEED=1 cython sklearn...` and see if it becomes consistent.

Wow nice, I did not think about that all I have to say! It does become consistent when setting PYTHONHASHSEED indeed.

Would it make sense to have a better default behaviour by having a consistent ordering though? Using OrderedDict for Python >= 2.7 jumps to mind.

I was wondering whether you supported Python 2.6, turns out you do https://github.com/cython/cython/wiki/FAQ#what-python-versions-does-cython-support.

**Comment by ogrisel** [comment]
Note that starting with Python 3.6, the dict type will become deterministic by default. Still having a deterministic dicts for Python 3.4 / 3.5  would be nice to be able to use ccache efficiently on our CI infrastructure.

**Comment by robertwb** [comment]
We still would rather produce deterministic files without relying on dict ordering or PYTHONHASHSEED. If you could cut this down to a minimal example, I could try and provide a fix.

**Comment by lesteve** [comment]
I cut it down to a simple example .pyx:

```pyx
# test.pyx
from cpython cimport array
cimport numpy as np
```

```
❯ (for i in $(seq 1 10); do cython test.pyx; md5sum test.c; done) | sort
690161681d498f1c9a7c75e24251117d  test.c
690161681d498f1c9a7c75e24251117d  test.c
690161681d498f1c9a7c75e24251117d  test.c
690161681d498f1c9a7c75e24251117d  test.c
690161681d498f1c9a7c75e24251117d  test.c
690161681d498f1c9a7c75e24251117d  test.c
f2e5b6ec698a9d0e90d9d48a6268d01f  test.c
f2e5b6ec698a9d0e90d9d48a6268d01f  test.c
f2e5b6ec698a9d0e90d9d48a6268d01f  test.c
f2e5b6ec698a9d0e90d9d48a6268d01f  test.c
```

**Comment by lesteve** [comment]
Note that may not be the only source of non-determinism since the diff does not look at all like the one I showed above.

**Comment by lesteve** [comment]
Another example I found.

```pyx
# test2.pxd
cdef inline int func():
    return 1
```

```pyx
# test2.pyx
from cpython cimport array
```

Output:
```
❯ (for i in $(seq 1 10); do cython test2.pyx; md5sum test2.c; done) | sort
2b629af3e2771b0638bbe10772529753  test2.c
2b629af3e2771b0638bbe10772529753  test2.c
2b629af3e2771b0638bbe10772529753  test2.c
2b629af3e2771b0638bbe10772529753  test2.c
2b629af3e2771b0638bbe10772529753  test2.c
2b629af3e2771b0638bbe10772529753  test2.c
dcceaf8d78f1c0540e553977cf6bf7e1  test2.c
dcceaf8d78f1c0540e553977cf6bf7e1  test2.c
dcceaf8d78f1c0540e553977cf6bf7e1  test2.c
dcceaf8d78f1c0540e553977cf6bf7e1  test2.c
```

 The diff looks closer to the one in the original issue:
```diff
❯ diff test2.c{1,2}
diff --git a/test2.c1 b/test2.c2
index 8c0caef..2ac4f90 100644
--- a/test2.c1
+++ b/test2.c2
@@ -988,6 +988,34 @@ static PyObject *__pyx_n_s_test;
 static int __pyx_pf_7cpython_5array_5array___getbuffer__(arrayobject *__pyx_v_self, Py_buffer *__pyx_v_info, CYTHON_UNUSED int __pyx_v_flags); /* proto */
 static void __pyx_pf_7cpython_5array_5array_2__releasebuffer__(CYTHON_UNUSED arrayobject *__pyx_v_self, Py_buffer *__pyx_v_info); /* proto */
 
+/* "test2.pxd":1
+ * cdef inline int func():             # <<<<<<<<<<<<<<
+ *     return 1
+ */
+
+static CYTHON_INLINE int __pyx_f_5test2_func(void) {
+  int __pyx_r;
+  __Pyx_RefNannyDeclarations
+  __Pyx_RefNannySetupContext("func", 0);
+
+  /* "test2.pxd":2
+ * cdef inline int func():
+ *     return 1             # <<<<<<<<<<<<<<
+ */
+  __pyx_r = 1;
+  goto __pyx_L0;
+
+  /* "test2.pxd":1
+ * cdef inline int func():             # <<<<<<<<<<<<<<
+ *     return 1
+ */
+
+  /* function exit code */
+  __pyx_L0:;
+  __Pyx_RefNannyFinishContext();
+  return __pyx_r;
+}
+
 /* "array.pxd":91
  *             __data_union data
  * 
@@ -1622,34 +1650,6 @@ static CYTHON_INLINE void __pyx_f_7cpython_5array_zero(arrayobject *__pyx_v_self
   __Pyx_RefNannyFinishContext();
 }
 
-/* "test2.pxd":1
- * cdef inline int func():             # <<<<<<<<<<<<<<
- *     return 1
- */
-
-static CYTHON_INLINE int __pyx_f_5test2_func(void) {
-  int __pyx_r;
-  __Pyx_RefNannyDeclarations
-  __Pyx_RefNannySetupContext("func", 0);
-
-  /* "test2.pxd":2
- * cdef inline int func():
- *     return 1             # <<<<<<<<<<<<<<
- */
-  __pyx_r = 1;
-  goto __pyx_L0;
-
-  /* "test2.pxd":1
- * cdef inline int func():             # <<<<<<<<<<<<<<
- *     return 1
- */
-
-  /* function exit code */
-  __pyx_L0:;
-  __Pyx_RefNannyFinishContext();
-  return __pyx_r;
-}
-
 static PyMethodDef __pyx_methods[] = {
   {0, 0, 0, 0}
 };
@@ -1809,9 +1809,12 @@ PyMODINIT_FUNC PyInit_test2(void)
   if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_1) < 0) __PYX_ERR(1, 1, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
 
-  /* "test2.pxd":1
- * cdef inline int func():             # <<<<<<<<<<<<<<
- *     return 1
+  /* "array.pxd":159
+ *     return extend_buffer(self, other.data.as_chars, Py_SIZE(other))
+ * 
+ * cdef inline void zero(array self):             # <<<<<<<<<<<<<<
+ *     """ set all elements of array to zero. """
+ *     memset(self.data.as_chars, 0, Py_SIZE(self) * self.ob_descr.itemsize)
  */
 
   /*--- Wrapped vars code ---*/
```

**Comment by scoder** [comment]
This is still a problem with 0.28.3.

**Comment by scoder** [comment]
I cannot reproduce this any more with 0.29, so I'll assume that this is fixed now.
